internal/k8s/statefulset/statefulset.go

```diff
@@ -0,0 +1,141 @@
+//
+// Copyright (C) 2019-2020 Vdaas.org Vald team ( kpango, rinx, kmrmt )
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+package statefulset
+
+import (
+	"context"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/vdaas/vald/internal/errors"
+	"github.com/vdaas/vald/internal/k8s"
+	appsv1 "k8s.io/api/apps/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// StatefulSetWatcher is a type alias for k8s resource controller
+type StatefulSetWatcher k8s.ResourceController
+
+type reconciler struct {
+	ctx         context.Context
+	mgr         manager.Manager
+	name        string
+	namespace   string
+	onError     func(err error)
+	onReconcile func(rs map[string][]StatefulSet)
+}
+
+// Statefulset is a type alias for the k8s statefulset definition.
+type StatefulSet = appsv1.StatefulSet
+
+// New returns the StatefulSetWather that implements reconciliation loop, or any error occured.
+func New(opts ...Option) (StatefulSetWatcher, error) {
+	r := new(reconciler)
+
+	for _, opt := range append(defaultOpts, opts...) {
+		if err := opt(r); err != nil {
+			return nil, errors.ErrOptionFailed(err, reflect.ValueOf(opt))
+		}
+	}
+
+	return r, nil
+}
+
+// Reconcile implements k8s reconciliation loop to retrive the StatefulSet information from k8s.
+func (r *reconciler) Reconcile(req reconcile.Request) (res reconcile.Result, err error) {
+	ssl := new(appsv1.StatefulSetList)
+
+	err = r.mgr.GetClient().List(r.ctx, ssl)
+	if err != nil {
+		if r.onError != nil {
+			r.onError(err)
+		}
+		res = reconcile.Result{
+			Requeue:      true,
+			RequeueAfter: time.Millisecond * 100,
+		}
+		if k8serrors.IsNotFound(err) {
+			res.RequeueAfter = time.Second
```

Comment:
```suggestion
			log.Error("not found", err)
			res.RequeueAfter = time.Second
			return res, err
```

Reply:
I think below is better.
what do you think about which is the best ?
@kevindiu @hlts2 
```suggestion
			log.Error("Not Found", err)
			res.RequeueAfter = time.Second
```

Reply:
how about this?
```suggestion
			log.Error(errors.Wrap(err, "not found"))
			res.RequeueAfter = time.Second
```

Reply:
🤔 
btw, which is general `not found` or `Not Found` in Vald?

Reply:
I think small case one is general in vald.
https://github.com/vdaas/vald/search?q=errors.New